pkg/alertmanager/multitenant_test.go

```diff
@@ -368,3 +395,456 @@ receivers:
+		},
+		{
+			name:              "sharding enabled, 3 instances, RF = 2",
+			withSharding:      true,
+			instances:         3,
+			replicationFactor: 2,
+			configs:           10,
+			expectedTenants:   20, // configs * replication factor
+		},
+		{
+			name:              "sharding enabled, 5 instance, RF = 3",
+			withSharding:      true,
+			instances:         5,
+			replicationFactor: 3,
+			configs:           10,
+			expectedTenants:   30, // configs * replication factor
+		},
+	}
+
+	for _, tt := range tc {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := context.Background()
+			ringStore := consul.NewInMemoryClient(ring.GetCodec())
+			mockStore := &mockAlertStore{
+				configs: map[string]alerts.AlertConfigDesc{},
+			}
+
+			var instances []*MultitenantAlertmanager
+			var instanceIDs []string
+			registries := util.NewUserRegistries()
+
+			// First, add the number of configs to the store.
+			for i := 1; i <= tt.configs; i++ {
+				u := fmt.Sprintf("u-%d", i)
+				mockStore.configs[u] = alerts.AlertConfigDesc{
+					User:      u,
+					RawConfig: simpleConfigOne,
+					Templates: []*alerts.TemplateDesc{},
+				}
+			}
+
+			// Then, create the alertmanager instances, start them and add their registries to the slice.
+			for i := 1; i <= tt.instances; i++ {
+				instanceIDs = append(instanceIDs, fmt.Sprintf("alertmanager-%d", i))
+				instanceID := fmt.Sprintf("alertmanager-%d", i)
+
+				amConfig := mockAlertmanagerConfig(t)
+				amConfig.ShardingRing.ReplicationFactor = tt.replicationFactor
+				amConfig.ShardingRing.InstanceID = instanceID
+				amConfig.ShardingRing.InstanceAddr = fmt.Sprintf("127.0.0.%d", i)
+				// Do not check the ring topology changes or poll in an interval in this test (we explicitly sync alertmanagers).
+				amConfig.PollInterval = time.Hour
+				amConfig.ShardingRing.RingCheckPeriod = time.Hour
+
+				if tt.withSharding {
+					amConfig.ShardingEnabled = true
+				}
+
+				reg := prometheus.NewPedanticRegistry()
+				am, err := createMultitenantAlertmanager(amConfig, nil, nil, mockStore, ringStore, log.NewNopLogger(), reg)
+				require.NoError(t, err)
+				defer services.StopAndAwaitTerminated(ctx, am) //nolint:errcheck
+
+				require.NoError(t, services.StartAndAwaitRunning(ctx, am))
+
+				instances = append(instances, am)
+				instanceIDs = append(instanceIDs, instanceID)
+				registries.AddUserRegistry(instanceID, reg)
+			}
+
+			// If we're testing sharding, we need make sure the ring is settled.
+			if tt.withSharding {
+				ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+				defer cancel()
+
+				// The alertmanager is ready to be tested once all instances are ACTIVE and the ring settles.
+				for _, am := range instances {
+					for _, id := range instanceIDs {
+						require.NoError(t, ring.WaitInstanceState(ctx, am.ring, id, ring.ACTIVE))
+					}
+				}
+			}
+
+			// Now that the ring has settled, sync configs with the instances.
+			var NumConfigs, NumInstances int
```

Comment:
No need to export these variables. Please name with with a lowercase first letter.